src/assign.rs

```diff
@@ -58,25 +54,456 @@ pub struct Assignment<'a> {
+    token: Token<'_>,
+    remaining: &Pointer,
+    mut buf: PointerBuf,
+    obj: &'v mut Map<String, Value>,
+    src: Value,
+) -> Result<Assigned<'v>, AssignError> {
+    // grabbing the entry of the token
+    let entry = obj.entry(token.to_string());
+    // adding token to the pointer buf
+    buf.push_back(token);
+
+    match entry {
+        Entry::Occupied(entry) => {
+            // if the entry exists, we either replace it or continue
+            let entry = entry.into_mut();
+            if remaining.is_root() {
+                // if this is the last token, we are done
+                // grab the old value and replace it with the new one
+                let replaced = Some(mem::replace(entry, src));
+                Ok(Assigned::Done(Assignment {
+                    assigned: entry,
+                    assigned_to: buf,
+                    replaced,
+                }))
+            } else {
+                // if this is not the last token, we continue with a mutable
+                // reference to the entry as the next value
+                Ok(Assigned::Continue {
+                    same_src: src,
+                    next_buf: buf,
+                    next_value: entry,
+                })
+            }
+        }
+        Entry::Vacant(entry) => {
+            // if the entry does not exist, we create a value based on the
+            // remaining path with the src value as a leaf and assign it to the
+            // entry
+            let src = expand_src_path(remaining, src)?;
+            let assigned = entry.insert(src);
+            Ok(Assigned::Done(Assignment {
+                assigned,
+                assigned_to: buf,
+                replaced: None,
+            }))
+        }
+    }
+}
+
+fn assign_scalar<'v>(
+    token: Token<'_>,
+    remaining: &'_ Pointer,
+    mut buf: PointerBuf,
+    value: &'v mut Value,
+    src: Value,
+) -> Result<Assigned<'v>, AssignError> {
+    // scalar values are always replaced at the current buf (with its token)
+    // build the new src and we replace the value with it.
+
+    buf.push_back(token);
+    let src = expand_src_path(remaining, src)?;
+    let replaced = Some(mem::replace(value, src));
+
+    Ok(Assigned::Done(Assignment {
+        assigned: value,
+        assigned_to: buf,
+        replaced,
+    }))
+}
+
+fn expand_src_path(mut path: &Pointer, mut src: Value) -> Result<Value, AssignError> {
+    while let Some((ptr, tok)) = path.split_back() {
+        path = ptr;
+        match tok.decoded().as_ref() {
+            "0" | "-" => {
+                src = Value::Array(vec![src]);
+            }
+            _ => {
+                let mut obj = Map::new();
+                obj.insert(tok.to_string(), src);
+                src = Value::Object(obj);
+            }
+        }
+    }
+    Ok(src)
+}
+
+#[cfg(test)]
+mod tests {
+    use serde_json::{json, Value};
+
+    use crate::{Pointer, PointerBuf};
+
+    #[test]
+    fn test_assign() {
```

Comment:
Probably also worth testing assigning directly to root (replacing it), both from an empty base object and a non-empty one.